File: Anya_Seth/Sliding Window/Sliding Window Maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
       int[] ans = new int[nums.length - k + 1];
       int[] leftMax = new int[nums.length];
       int[] rightMax = new int[nums.length];
       leftMax[0] = nums[0];
       for(int i = 1 ; i < nums.length ; i++){
          if(i % k  == 0){
            leftMax[i] = nums[i];
          }
          else{
          leftMax[i] = Math.max(leftMax[i-1] , nums[i]);
          }
       }
       rightMax[nums.length - 1] = nums[nums.length - 1];
       for(int i = nums.length - 2 ; i >= 0 ; i--){
          if((i+1) % k  == 0){
            rightMax[i] = nums[i];
          }
          else{
          rightMax[i] = Math.max(rightMax[i + 1], nums[i]);
          }
        } 
       for(int i = 0 ; i < ans.length ; i++){
          ans[i] = Math.max(leftMax[i + k - 1], rightMax[i]);
        }
        return ans;
    }
}

File: Ashutosh-Singh/SlidingWindow/Sliding-Window-Maximum.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> leftPrefixMax(n), rightPrefixMax(n), ans(n - k + 1);

        for (int i = 0; i < n; i++) {
            if (i % k == 0) {
                leftPrefixMax[i] = nums[i];
            } else {
                leftPrefixMax[i] = max(leftPrefixMax[i - 1], nums[i]);
            }
        }

        for (int i = n - 1; i >= 0; i--) {
            if (i == n - 1 || i % k == k - 1) {
                rightPrefixMax[i] = nums[i];
            } else {
                rightPrefixMax[i] = max(rightPrefixMax[i + 1], nums[i]);
            }
        }

        for (int i = 0; i < n - k + 1; i++) {
            ans[i] = max(rightPrefixMax[i], leftPrefixMax[i + k - 1]);
        }
        return ans;
    }
};

int main() {
    int n, k;
    cin >> n >> k;

    vector<int> nums(n);
    for (int i = 0; i < n; ++i) {
        cin >> nums[i];
    }

    Solution solution;
    vector<int> result = solution.maxSlidingWindow(nums, k);

    for (int val : result) {
        cout << val << " ";
    }
    cout << endl;

    return 0;
}
